# `anchovy` 

```
anchovy ><>
><>`
    ><>
```

![anchovies](https://github.com/QVEU/anchovy/blob/main/assets/northern-anchovies-rw07-130.webp)


## Overview

Pipeline to generate consensus viral sequences within individual cells in single-cell experiments. The package was developed for use with longread data (nanopore or PacBio). Scripts were developed to run on the NIH skyline cluster, and the job shell scripts are specific to packages installed there. Much work can be done to` improve its portability and reproducibility. 

## Barcode lists: 
- Barcode inclusion lists (FKA whitelists) are located in the 10X CellRanger `cellranger` installation directory (go to: `/path/to/cellranger-8.0.1/lib/python/cellranger/barcodes/`), or can be downloaded from various locations.
For example:
- v3 chemistry, `3M`, `3M-february-2018.txt.gz`: (https://teichlab.github.io/scg_lib_structs/data/10X-Genomics/3M-february-2018.txt.gz)[https://teichlab.github.io/scg_lib_structs/data/10X-Genomics/3M-february-2018.txt.gz]
- v4 chemistry, 16-bp cell barcodes can be found in `3M-3pgex-may-2023.txt.gz`: (https://teichlab.github.io/scg_lib_structs/data/10X-Genomics/3M-3pgex-may-2023.txt.gz)[https://teichlab.github.io/scg_lib_structs/data/10X-Genomics/3M-3pgex-may-2023.txt.gz]

## `anchovy` Functions
1. The script `anchovy.py` processes a mapped SAM file, usually generated by running `minimap2` against a viral reference. The script identifies the 10X barcode signature, reads and collects the cell barcodes (CBC) based on the whitelist, and collects read sequence, barcode position, from each individual mapped read into a csv, the 'anchovy csv'.
2. `UMItoFasta.py` converts reads in `anchovy.csv` to a collection of fasta files, one for each CBC.
3. These fastas are converted to a sam file, again by `minimap2`
4. `Sam2Consensus.py` is a little script used to generate the consensus per cell from the CBC-specific SAM files.
5. The fastas are combined using `cat` to yield the `allConsensus.fasta`
6. `ConsensusTool.py` (https://github.com/QVEU/anchovy/edit/main/README.md#consensustoolpy) computes a consensus sequence for each collection of reads in the CBC fasta file. python script processes these consensus sequences to extract reads corresponding to the region of interest, usually the viral ORF. This outputs a consensus sequence of all the cells and the `filtConsensus.csv`.
7. Accessory R script, `Consensus_Annotation.R` is used to convert these consensus sequences into the cell genotypes, and generate the genotype network from the `filtConsensus.csv`

- Steps 1-5 are included in the AnchovyJob.sh scripts in this repository, and instructions below. 

## Prerequisites
To run anchovy (steps 1-5)
- Python 3.8+
- sam2consensus.py - from [https://github.com/edgardomortiz/sam2consensus](https://github.com/edgardomortiz/sam2consensus)

To run `ConsensusTools`
- ipynb

To run `Consensus_Annotation.R`
- R 13.1+
- Rstudio
- ggplot2

## Installing
The scripts are all on the QVEU skyline space at `/data/lvd_qve/QVEU_code/Anchovy/`

To install locally:

1. Clone the repository:
    ```bash
    gh repo clone QVEU/anchovy
    git clone https://github.com/QVEU/anchovy.git
    ```
    
2. Navigate to the project directory:
    ```bash
    cd anchovy
    ```
    
## Usage

### Running Anchovy

There's two ways to run the `anchovy` code: 

- One is to run an edited version of the `AnchovyJob.sh` file. There are two examples in the repository, one for pacbio "PB" and one for Nanopore "NP", PB is most common.
- This will output several files, including:


### Anchovy Job run

Interactively:
```bash

sh AnchovyJob_script.sh path/to/whitelist_file path/to/inputdirectory path/to/template.fasta path/to/inputfilename.sam <platform>

```

Job Submission with Slurm:
```bash

sbatch AnchovyJob_script.sh path/to/whitelist_file path/to/inputdirectory path/to/template.fasta path/to/inputfilename.sam <platform>

```



### Post-analysis 

#### `ConsensusTool.py`
- Filters the consensus seuences from each cell to cover a specfic region, usually a coding region, ORF. Uses `...allConsensus.fasta` file from `anchovy.py` as input.

```python ConsensusTool.py <NAME>_allConsensus.fasta <ORFstartNT> <ORFendNT>```

Outputs: 
- `<NAME>_consensus_reference.txt`
- `<NAME>_filtConsensus.csv`
- `<NAME>_filtConsensus.fasta`

For example: 
```
python ConsensusTool.py /Volumes/LVD_qve/Projects/DENV_SEARCHLIGHT/Anchovy_2/DENV_6dpi_allConsensus.fasta 96 10272
```

#### Analysis with `Consensus_Annotation.R` R script
  
  - Translation and interpretation of mutations, generation of genotype network. 
```bash
    Rscript ~/path/to/anchovy/Consensus_Annotation_v3.R /<filtConsensus_reference.txt> <filtConsensus.csv> /path/to/output/file/<prefix>
```

For example:
```bash
    Rscript ~/Documents/GitHub/anchovy/Consensus_Annotation.R ~/reference_consensus.txt> <filtConsensus.csv> /path/to/output/file/<prefix>
```
Output: 
- `<prefix>.pdf`
- `<prefix>_annot_v3.csv`
- `<prefix>_epistaticNetwork.csv`
- `<prefix>_genotypeNetwork.csv`



### Contributing

- Fork the repository.
- Create a new branch (git checkout -b feature-branch).
- Commit your changes (git commit -am 'Add new feature').
- Push to the branch (git push origin feature-branch).
- Open a pull request.
